src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelModuleResolutionFunction.java

```diff
@@ -259,14 +266,41 @@ private static ImmutableMap<ModuleKey, Module> computeFinalDepGraph(
       ImmutableMap<String, ModuleOverride> overrides,
       ExtendedEventHandler eventHandler)
       throws BazelModuleResolutionFunctionException, InterruptedException {
-    ImmutableMap.Builder<ModuleKey, Module> finalDepGraph = ImmutableMap.builder();
-    for (Map.Entry<ModuleKey, InterimModule> entry : resolvedDepGraph.entrySet()) {
-      finalDepGraph.put(
-          entry.getKey(),
-          moduleFromInterimModule(
-              entry.getValue(), overrides.get(entry.getKey().getName()), eventHandler));
+    ExecutorService executorService = Executors.newWorkStealingPool(8);
```

Comment:
First of all I'll admit I'm very unfamiliar with the Java parallel stream stuff, so apologies if I make any rookie mistakes here :)

From my read of things, it looks like we don't even need to have this executor service at all? We could just directly call `collect` on a parallel stream and Java would use the global fork join pool (`ForkJoinPool.commonPool()`) underneath. That's probably better than using a dedicated pool (as the global fork join pool is also used to carry virtual threads).

But I'm not sure what happens when an exception is thrown in that case. In sequential streams, I think the exception just gets propagated out, so we just need to "tunnel" it. Do you know what happens in parallel streams?

Reply:
The problem with using the common pool is that it has `NUM_CPUS - 1` threads by default, which can interact badly with other parallel operations relying on it, especially since we are blocking on IO.

Exceptions in parallel streams are also thrown by the terminal operation. I think that the first throw "wins", but I'm not sure about that.

Reply:
Thanks for the explanation. I did some more reading on parallel streams and I increasingly feel like they're not a great fit for our use case (maybe I just don't understand them well enough). My main gripe is the fact that it tries to do all the MapReduce-esque splitting and merging which is basically completely unnecessary here where we're just doing a map and basically no reduce. Plus the concurrency primitives are completely hidden and it works "magically" depending on whether you're already in a ForkJoinPool or not; and this doesn't seem to be documented anywhere in the official JDK -- I had to believe StackOverflow. (For example, I'm not confident at all whether the introduction of Loom would automatically cause parallel stream processing to use virtual threads.)

On the other hand, I'll admit that the magic isn't all bad -- just slapping `.parallel()` on the stream and having it work is pretty neat.

I'd probably have preferred to use Skyframe for this; that is, create a RepoSpecFunction/Value and just give everything to Skyframe. To avoid the restart, we could use a SkyKeyComputeState to store everything right before the repo spec computation step. WDYT?

Reply:
cc @lberki

Reply:
Yeah, the lack of official documentation on parallel streams' internals really hurts their usefulness. I can only hope that this explained by the desire to switch them over to a virtual thread executor by default once available - one can dream.

I don't quite see the problem with us not using the full feature set of a MapReduce. Isn't turning a list of values into a `LinkedHashMap` still an (associative) reduction? But we do need to believe StackOverflow that streams are just relying on `ForkJoinPool#fork` and thus just depens on the `ForkJoinPool` implementation that we started with (`#commonPool` or a custom one).

I am fine with switching to Skyframe here though, especially as it avoids anyone reading the code from having to reason about another concurrency framework.

Reply:
I'm not familiar with the code here so I can't have a fully educated opinion, but in general, the framework we use in Blaze to schedule computation is Skyframe. What is the argument against using it?

I can imagine that:
* It's too memory-intensive to keep the interim results of the computation around forever
* There isn't a convenient "name" for any piece of computation that could be used as a `SkyKey`

Notably, @Wyverald's recent change to keep Starlark threads around did use a separate thread pool, but in that case, threads are really used as a way to keep track of the state of the computation over Skyframe restarts, which is an acceptable use case (until Loom takes off, anyway)

Reply:
I think I found a way to extract this work into a SkyFunction, please take a look. I am a bit unsure about how to implement `Registry`'s equality correctly, which made me worried about using it in a `SkyKey`.

Reply:
I trimmed this down to a `ModuleKey` and a `String`.

Reply:
Yep, that sounds much nicer. I'll leave the review to @Wyverald since he is much more familiar with the area (unless he indicates that he could use my brain)

Reply:
> I don't quite see the problem with us not using the full feature set of a MapReduce. Isn't turning a list of values into a LinkedHashMap still an (associative) reduction?

That was poorly worded on my part -- what I wanted to say was that the extra considerations about splitting the stream into appropriately sized substreams so that it could be sent to child ForkJoinTasks and then merging the results together are kind of wasteful. Yes technically a "map" operation is a "reduce" operation itself too, but we really don't need a tree structure for this. Just linearly going over the keys and sending each one off for a download will suffice.

So I'm very happy with the new Skyframe-based implementation :)